### ISSUE TITLE
Pass argument of struct type to call_state()

### ISSUE THREAD

**Issue by tbbatbb** [angr/angr#2583]
Hi, I want to make a calling state with an argument of struct type by calling `proj.factory.call_state`, but `angr` seems to replace the **argument** with **the address of the argument** before the calling. Here is the code:

```c
typedef struct _A {
    int a;
    char b;
    short c;
    void* d;
    int e;
} A;

void func(A a) {
    // ...
}
```
I want to pass an argument with values `(a=10, b=11, c=12, d=0x66666666, e=13)` to `func`. And the following code I used didn't work as I expected:

```python
import angr

# loading binary
proj = angr.Project('bins/exe', load_options={
    'auto_load_libs': False,
    'main_opts': {
        'base_addr': 0
    }
})

# getting rebased address of func
func_addr = proj.loader.find_symbol('func').rebased_addr

# function definition
ty = angr.types.parse_defns('void foo(struct a {int a; char b; short c; void* d; int e; } x);')['foo']

cc = proj.factory.cc(ty=ty)

s = proj.factory.call_state(func_addr, (10, 11, 12, 0x66666666, 13), cc=cc, ret_addr=0x12345678)
```
and when I call `s.stack_read`, I got following data:

```python
s.regs.esp               # 0x7ffeffe9

s.stack_read(0, 4)   # 0x12345678    ret_addr

s.stack_read(4, 4)   # 0x7ffefff1        the address of struct argument

s.stack_read(8, 4)   # 10                    a.a

s.stack_read(12, 4)   # 11                  a.b
```

but when I compile the source code with `gcc 7.5.0` on 'unbuntu 18.04 x86`, and I make a `breakpoint` in `func` to inspect the stack, then I got this

![image](https://user-images.githubusercontent.com/20682299/113085493-8007e600-9212-11eb-91a1-34fa0df37cd4.png)

with nothing between `ret_addr` and the start of argument `a`.

There must be some mistakes I made while calling `proj.factory.call_state`, but I can't figure out.

**Comment by rhelmot** [comment]
Our calling conventions don't support serializing structs this way :( this has been a longstanding TODO that nobody has needed up until now.

**Comment by tbbatbb** [comment]
Thanks for your reply. I searched the doc and google it for a long time, however, I haven't found any solution. It works when I use `proj.factory.blank_state` and pushing arguments into stack manually in this case, but in other cases, arguments can be passed with registers, then I have no idea of how to deal with it. It seems that `angr` automatically convert the **struct** into **a pointer of that struct** when a argument of that type passed into function. Is there any possibility that I can pass an argument of structs into function directly? Or maybe there is some APIs in `angr` that it can tell me how to pass arguments when calling function, for example, when the function has a `usercall` calling convention, I can get the information about how arguments supposed to be passed with specific register?

**Comment by rhelmot** [comment]
yes, angr will automatically convert structs into pointers to structs. Not doing this is a missing feature, and I don't know when I will be able to find the time to fix it.

**Comment by tbbatbb** [comment]
OK, thanks a lot anyway.

**Comment by tbbatbb** [comment]
Well, I found that `angr` do the same to arguments with type `long long`, it converts **`long long`** to **pointers to `long long`** as well,  while it can handle `double` correctly, confusing. 

Source code:
```c
void func(char a, short b, int c, long long d, double e) {
    // ...
}
```

Python with `angr`:
```python
p = angr.Project('exe', load_options={
     'auto_load_libs': False,
     'main_opts': {
         'base_addr': 0
     }
})

ty = angr.types.parse_defns('void foo(char a, short b, int c, long long d, double e);')['foo'] 

s = p.factory.call_state(0x567, 1,2,3,4,5, cc=p.factory.cc(func_ty=ty), ret_addr=0x66666666)

s.stack_read(0,4)     # ret_addr   <BV32 0x66666666>

s.stack_read(16,4)   # addr of d  <BV32 0x7ffefff8>

# as if there is not argument d
ty = angr.types.parse_defns('void foo(char a, short b, int c, double e);')['foo']              

s = p.factory.call_state(0x567, 1,2,3,4, cc=p.factory.cc(func_ty=ty), ret_addr=0x66666666)   

s.stack_read(16,8)   # e    <BV64 0x4010000000000000>
```

**Comment by rhelmot** [comment]
this has finally been resolved! see #2961. You may now use a dict (or a tuple, or a SimStructValue) to pass structs by value, provided you follow the new guidelines for providing function prototypes to call_state.

**Comment by tbbatbb** [comment]
您好，感谢您的来信，您的邮件我已收到。我将尽快给您回复。
